pkgs/top-level/config.nix

```diff
@@ -151,12 +235,39 @@ in {
         in r // { _undeclared = r; };
     };
 
+  imports = [
+    (mkRemovedOptionModule ["config" "showDerivationWarnings"] "See `config.problems.matchers` and the respective documentation. Be aware that the semantics have changed!")
+  ];
+
+
   inherit options;
 
   config = {
     warnings = lib.optionals config.warnUndeclaredOptions (
       lib.mapAttrsToList (k: v: "undeclared Nixpkgs option set: config.${k}") config._undeclared or {}
     );
+
+    assertions =
+      # Collect the assertions from the problems.matchers.* submodules, propagate them into here
+      lib.concatMap (matcher: matcher.assertions) config.problems.matchers;
+
+    # Put the default value for matchers in here (as in, not as an *actual* mkDefault default value),
+    # to force it being merged with any custom values instead of being overridden.
+    problems.matchers = [
+      # Be loud and clear about package removals
+      {
+        kind = "removal";
+        handler = "error";
+      }
+      {
+        kind = "maintainerless";
+        handler = "ignore";
+      }
+      # Wildcard for everything else
+      {
+        handler = "ignore";
+      }
```

Comment:
Won't this shadow whatever the user adds into `problem.matchers`, unless they use `mkForce`?

Reply:
The current idea is to have everything merged into a list, since it will take the highest severity handler if multiple apply this should be a no-op.